### ISSUE TITLE
Error "all dependencies must come from the same source" is unclear

### ISSUE THREAD

**Issue by Susurrus** [rust-lang/cargo#2947]
When running `cargo publish` on my project after following the steps outlined [in this guide](http://doc.crates.io/crates-io.html), I got the `error: all dependencies must come from the same source". The repo I'm trying to push is on [Gitlab](https://gitlab.com/susurrus/gatii).

Now given this error and the guide on crates.io I was following, I'm stuck. The error isn't clear enough and the verbose mode of `cargo publish` didn't provide any more details for me to resolve the issue. More details on what exactly the dependency requirements are and why they have failed would be good. I think it'd also be nice if this was part of the `cargo package` command rather than the `cargo publish` as I would've thought if `cargo package` succeeds, everything is good-to-go for pushing to crates.io.

**Comment by alexcrichton** [comment]
Certainly! Sounds like a good starter bug for anyone who's interested, should be easy to expand the information and update the tests here as well.

**Comment by Susurrus** [comment]
@alexcrichton That would definitely be nice, and I think I could contribute there, but this is blocking me publishing my crate to crates.io. Could you point me to where the docs describe this issue or speak to it if you know it personally? I'd like to get my crate fixed and published and once I understand the issue enough to fix it, I'll come back and tackle this issue.

**Comment by alexcrichton** [comment]
Oh sorry, right! This happens if you have, for example, git dependencies. Crates published to crates.io are only allowed to depend on crates from crates.io.

**Comment by Susurrus** [comment]
Thanks Alex.

Hmm, I would not have guessed that but it's probably a signature guarantee thing. Is this documented anywhere that's not in the cargo codebase? I liked to the [Crates Publishing Guide](http://doc.crates.io/crates-io.html) and it doesn't mention it (I can't find it by searching for 'dependenc'). I think I should start with that and then move on to modifying Cargo to spit out a better error message.

**Comment by alexcrichton** [comment]
Ah no I think it unfortunately may not be documented right now, but I think it'd be a great idea to update the docs!

**Comment by Susurrus** [comment]
I'm happy to do it assuming I can get the full scoop on the actual requirements (which I can write up in some nicer English) and I can get pointed in the right direction. Is the crates.io document I linked to the right location to put this and I should create an issue there about this?

**Comment by alexcrichton** [comment]
Sure! That documentation actually lives in this repo at `src/doc`, so feel free to send a PR here as well.

**Comment by Susurrus** [comment]
So I've been working on fixing this, but there are actually a couple different errors you can get, both of which are not helpful in resolving the error, just pointing it out:

1. If you have a dependency that is a `path`, then you will get the error:
```shell
$ cargo publish --dry-run --allow-dirty
Updating registry `https://github.com/rust-lang/crates.io-index`
error: all path dependencies must have a version specified when publishing.
dependency `CRATE` does not specify a version
```
2. If you have a dependency that is a `git` repository then you will get the error:
```shell
Updating registry `https://github.com/rust-lang/crates.io-index`
error: all dependencies must come from the same source.
dependency `CRATE` comes from REPO_URL instead
```

When running `cargo publish` I believe the ONLY acceptable source is a version (don't know how this is referred to internally), though the first error implies otherwise. I think the error messages for both of these should be the same. I'm not sure what exactly it should look like cause I don't know what the only allowable source should be referred to as. I think this can only allow crates.io currently, so maybe we can just say the following and update it once federation is supported:

1. For a dependency on a path:
```shell
Updating registry `https://github.com/rust-lang/crates.io-index`
error: All dependencies must be obtainable from crates.io.
dependency `CRATE` comes from local path "LOCAL_PATH" instead
```
2. For a dependency on a repository (currently just `git`):
```shell
Updating registry `https://github.com/rust-lang/crates.io-index`
error: All dependencies must be obtainable from crates.io.
dependency `CRATE` comes from REPO_TYPE repository at "REPO_URL" instead
```

**Comment by onur** [comment]
For the first one; all crates uploaded to crates.io must specify a "version" for dependencies. Even if you are using a path dependency, you need to specify its version. This is documented in cargo docs: <http://doc.crates.io/specifying-dependencies.html#specifying-path-dependencies>

**Comment by Susurrus** [comment]
So it looks like you can specify a path but you must manually assign it a version (though I've never tried to do this). I assume this is primarily used for multiple crates that are tracked in the same repo, yes?

So it looks like point 2. is all that's necessary to change.

**Comment by alexcrichton** [comment]
Sounds like a reasonable change to me!

**Comment by Susurrus** [comment]
Resolved in #3606.
